### DISCUSSION TITLE
Simulation time issue

### DISCUSSION THREAD

**Discussion by ahmetozleraktas** [lava-nc/lava#272] [Q&A]
Hello everyone, I hope you're ok. I have an issue about simulation time of a tutorial. I changed a little bit of tutorial_2 Processes to achieve Izhikevich model. It takes too long to compute and I did not get any results for now. It also takes too long when I simulate without changing the LIF model in tutorial. I added the codes I used on Izhikevich model. What is the problem?

```
import numpy as np
from lava.magma.core.process.process import AbstractProcess
from lava.magma.core.process.variable import Var
from lava.magma.core.process.ports.ports import InPort, OutPort

class Izhikevich(AbstractProcess):
    def __init__(self, **kwargs):
        super().__init__()
        shape = kwargs.get("shape", (1,))
        self.a_in = InPort(shape=shape)
        self.s_out = OutPort(shape=shape)
        self.u = Var(shape=shape, init=0)
        self.v = Var(shape=shape, init=0)
        self.du = Var(shape=(1,), init=kwargs.pop("du", 0))
        self.dv = Var(shape=(1,), init=kwargs.pop("dv", 0))
        self.a = Var(shape=(1,), init=kwargs.pop("a", 0.02))
        self.b = Var(shape=(1,), init=kwargs.pop("b", 0.2))
        self.c = Var(shape=(1,), init=kwargs.pop("c", -65))
        self.d = Var(shape=(1,), init=kwargs.pop("d", 2))
        self.I = Var(shape=shape, init=kwargs.pop("I", 0))
        self.vth = Var(shape=(1,), init=kwargs.pop("vth", 30))

    def print_vars(self):
        """Prints all variables of a LIF process and their values."""

        sp = 3 * "  "
        print("Variables of the Izhikevich:")
        print(sp + "u:    {}".format(str(self.u.get())))
        print(sp + "v:    {}".format(str(self.v.get())))
        print(sp + "du:   {}".format(str(self.du.get())))
        print(sp + "dv:   {}".format(str(self.dv.get())))
        print(sp + "a:    {}".format(str(self.a.get())))
        print(sp + "b:    {}".format(str(self.b.get())))
        print(sp + "c:    {}".format(str(self.c.get())))
        print(sp + "d:    {}".format(str(self.d.get())))
        print(sp + "I:    {}".format(str(self.I.get())))
        print(sp + "vth:  {}".format(str(self.vth.get())))
```

```
import numpy as np
from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
from lava.magma.core.model.py.ports import PyInPort, PyOutPort
from lava.magma.core.model.py.type import LavaPyType
from lava.magma.core.resources import CPU
from lava.magma.core.decorator import implements, requires, tag
from lava.magma.core.model.py.model import PyLoihiProcessModel

@implements(proc=Izhikevich,protocol=LoihiProtocol)
@requires(CPU)
@tag('floating_pt')
class PyLifModel(PyLoihiProcessModel):
    a_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, float)
    s_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, bool, precision=1)
    u: np.ndarray = LavaPyType(float, float)
    v: np.ndarray = LavaPyType(float, float)
    du: float = LavaPyType(float, float)
    dv: float = LavaPyType(float, float)
    a: float = LavaPyType(float, float)
    b: float = LavaPyType(float, float)
    c: float = LavaPyType(float, float)
    d: float = LavaPyType(float, float)
    I: np.ndarray = LavaPyType(float, float)
    vth: float = LavaPyType(float, float)

    def run_spk(self):
        a_in_data = self.a_in.recv()
        self.dv = 0.04*self.v**2 + 5*self.v + 140 - self.u + self.I
        self.du = self.a*(self.b*self.v - self.u)
        s_out = self.v >= self.vth
        self.v[s_out] = self.c  # Reset voltage to c value
        self.s_out.send(s_out)
```

```
izhikev = Izhikevich(shape=(1,), du=0, dv=0, I=5, vth=30)

from lava.magma.core.run_configs import Loihi1SimCfg
from lava.magma.core.run_conditions import RunSteps


izhikev.run(condition=RunSteps(num_steps=1), run_cfg=Loihi1SimCfg())
izhikev.pause()
izhikev.print_vars()
```

**Comment by ahmetozleraktas** [comment]
So, codes not worked for me on jupyter notebook. I created a python file and tried same codes, it gaved another error. It solved with this code block and some edits on behavior model:

```
if __name__ == '__main__':
    lif.run(condition=RunSteps(num_steps=15), run_cfg=Loihi1SimCfg())
    lif.print_vars()
    lif.stop()
```

**Reply by GaboFGuerra** [reply]
Hi Ahmet, 

Thanks for filing the issue, may I ask if you meant the "codes **did not** work on Jupyter", or "codes **now** worked on Jupyter"?

**Reply by ahmetozleraktas** [reply]
Codes are still not working on jupyter. I am working on a .py file and that code block worked on .py file. Sorry for the uncertainty.

**Reply by GaboFGuerra** [reply]
Hi Ahmet,  

When you declare the LavaPyType for the variables, where you are using LavaPyType(float, float) you should be using something like LavaPyType(np.ndarray, np.float, precision=24), please see the signature for LavaPyType. Also, you are missing an izhikev.stop(), if you don't stop the process it will be just hanging waiting for instructions after pause.

**Reply by GaboFGuerra** [reply]
Regarding the LavaPyType, the first argument is used to create the model for the variable at runtime (the actual object implementation, in contrast to the Var at the process level), variables in general are array-like, so float would only work for constants not variables. The second argument is the type for every element of the variable's array and the third element is their precision. I now see we are missing a docstring for that dataclass.

**Reply by ahmetozleraktas** [reply]
I thought "float" is needed for a single value (not matrix), after all it would be correct since I would try it for a single neuron, but i fixed that later as you suggested. I somehow managed to simulate one izhikevich neuron and monitored the outputs for that, but runtime is very slow for 10k steps. Is that normal? I must point out: I'm running that simulation code as the accepted answer in this discussion with `izhikev.stop()` at the end.

**Reply by ahmetozleraktas** [reply]
Also, tutorials are not working on jupyter at all. Why is that happening?

**Comment by ahmetozleraktas** [comment]
**Last edit:**  The code part I mentioned works but it is really slow and I think it should not be take that long. I couldn't solve it on jupyter too. How did you run the tutorials without editing the notebooks?
